### PR TITLE
Package support

### PR DESCRIPTION
**Pull request by haifengkao**
Add `npm run pack` to build nyaovim.app.
Add `npm run dist` to build the corresponding dmg file.

### What was a problem?

Nyaovim didn't have the script to build the binary distribution file.

### How this PR fixes the problem?
I use  `electron-builder` to package the project.

### Check lists (check `x` in `[ ]` of list items)

- [x] Coding style (if any code was modified)
- [x] Confirmed
  - OS: OSX 10.11
  - `nvim` version: 0.1.5

### Additional Comments (if any)

I love you :)

### CODE REVIEW

File: main/main.ts [hunk 1]
```diff
@@ -135,6 +135,13 @@ app.on('open-url', (e: Event, u: string) => {
     e.preventDefault();
     shell.openExternal(u);
 });
+// we use once here because only the first open-file event might be missed by nyaovim-app
+app.once('open-file', (e: Event, p: string) => {
+    // open-file event might be sent before ready event is emitted
+    // put it in argv to let nyaovim-app to pick it up later
+    process.argv.push(p);
+    e.preventDefault();
+});
```
Comment (rhysd) [inline main/main.ts]: Nice catch!  But it looks not related to this PR directly.  Could you create another PR for this to manage history of development better?

File: package.json [hunk 2]
```diff
@@ -21,6 +23,27 @@
     "type": "git",
     "url": "git+https://github.com/rhysd/NyaoVim.git"
   },
+  "build": {
+    "asar": false,
+    "appId": "org.vim.NyaoVim",
```
Comment (rhysd) [inline package.json]: I think `neovim` is better than `vim` because Neovim is a different software from Vim.

File: package.json [hunk 3]
```diff
@@ -50,6 +72,8 @@
   "devDependencies": {
     "bower": "^1.7.1",
     "tslint": "^3.14.0",
+    "electron-builder": "^5.22.2",
+    "electron-prebuilt": "^1.3.2",
```
Comment (rhysd) [inline package.json]: Please do not move electron-prebuilt from dependency to dev dependency.  npm package won't work if Electron binary is in dev dependency.

File: renderer/nyaovim-app.ts [hunk 4]
```diff
@@ -205,8 +205,14 @@ Polymer({
         argv: {
             type: Array,
             value: function() {
-                // Note: First and second arguments are related to Electron
-                const a = remote.process.argv.slice(2);
+                var electron_argc =  1; // the first argument of standalone distribution is the application path
```
Comment (rhysd) [inline renderer/nyaovim-app.ts]: Please use `let` or `const` instead of `var` consistently.

File: main/main.ts (hunk 1 above)
Comment (haifengkao) [inline main/main.ts]: True. But this issue only happens when a user double-clicks a document to open NyaoVim. It cannot be reproduced for an npm module.

File: package.json (hunk 2 above)
Comment (haifengkao) [inline package.json]: agree

File: package.json (hunk 3 above)
Comment (haifengkao) [inline package.json]: This is a big issue. electron-builder asks me to move electron-prebuilt to dev dependency. Maybe we need a two different package.json?

File: renderer/nyaovim-app.ts (hunk 4 above)
Comment (haifengkao) [inline renderer/nyaovim-app.ts]: ok

File: package.json (hunk 3 above)
Comment (rhysd) [inline package.json]: I think it's better to execute `npm uninstall --save electron-prebuilt` before running builder then `git checkout package.json` after.  Although just a workaround.

File: main/main.ts (hunk 1 above)
Comment (rhysd) [inline main/main.ts]: Hmm... OK, so let's include this in this PR.

File: package.json [hunk 5]
```diff
@@ -21,6 +23,27 @@
     "type": "git",
     "url": "git+https://github.com/rhysd/NyaoVim.git"
   },
+  "build": {
+    "asar": false,
```
Comment (rhysd) [inline package.json]: Why packaging with asar is disabled?

Comment (haifengkao) [inline package.json]: easier for me to debug the code XD

Comment (rhysd) [inline package.json]: :)

File: main/main.ts [hunk 6]
```diff
@@ -135,6 +135,13 @@ app.on('open-url', (e: Event, u: string) => {
     e.preventDefault();
     shell.openExternal(u);
 });
+// we use once here because only the first open-file event might be missed by nyaovim-app
+app.once('open-file', (e: Event, p: string) => {
```
Comment (rhysd) [inline main/main.ts]: I think types for callback parameters don't need:

```typescript
app.once('open-file', (e, p) => {
```

**Comment by haifengkao** [comment]
superceded by https://github.com/rhysd/NyaoVim/pull/93
